service/cachemdw/cache.go

```diff
@@ -0,0 +1,171 @@
+package cachemdw
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/kava-labs/kava-proxy-service/clients/cache"
+	"github.com/kava-labs/kava-proxy-service/decode"
+	"github.com/kava-labs/kava-proxy-service/logging"
+)
+
+type EVMClient interface {
+	decode.EVMClient
+
+	ChainID(ctx context.Context) (*big.Int, error)
+}
+
+type ServiceCache struct {
+	cacheClient              cache.Cache
+	evmClient                EVMClient
+	cacheTTL                 time.Duration
+	decodedRequestContextKey any
+
+	*logging.ServiceLogger
+}
+
+func NewServiceCache(
+	cacheClient cache.Cache,
+	evmClient EVMClient,
+	cacheTTL time.Duration,
+	decodedRequestContextKey any,
+	logger *logging.ServiceLogger,
+) *ServiceCache {
+	return &ServiceCache{
+		cacheClient:              cacheClient,
+		evmClient:                evmClient,
+		cacheTTL:                 cacheTTL,
+		decodedRequestContextKey: decodedRequestContextKey,
+		ServiceLogger:            logger,
+	}
+}
+
+func IsCacheable(ctx context.Context, evmClient EVMClient, req *decode.EVMRPCRequestEnvelope) bool {
+	blockNumber, err := req.ExtractBlockNumberFromEVMRPCRequest(ctx, evmClient)
+	if err != nil {
+		return false
+	}
+
+	if blockNumber <= 0 {
+		return false
+	}
```

Comment:
nit: the decoding uses `-3` to signify "earliest" which should also be cacheable
https://github.com/Kava-Labs/kava-proxy-service/blob/69b86f0b542973522e729d8633d3291fb7bac192/decode/evm_rpc.go#L109-L113

not sure if it's worth capturing here.

for my work i was considering writing something that converted all the special block tags into true numbers, as for our purposes, `latest, pending, finalized, safe` are always "latest" (`0`) and `earliest` is always 1

@galxy25 it's true that "earliest" should always be block 1, yes? or do we only want to return the first block for which the node has data?

Reply:
to start we can assume earliest is block 1, but I can see this needing to be defined via environment variable at run time (based on the fact that for blockscout we configure indexing to start at block 100)

Reply:
will be fixed in follow-up PR